Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoGroupingExtensionRule.swift

```diff
@@ -22,38 +23,100 @@ struct NoGroupingExtensionRule: OptInRule {
     )
 
     func validate(file: SwiftLintFile) -> [StyleViolation] {
-        let collector = NamespaceCollector(dictionary: file.structureDictionary)
-        let elements = collector.findAllElements(of: [.class, .enum, .struct, .extension])
+        return Visitor(configuration: configuration, file: file)
+            .walk(tree: file.syntaxTree) { visitor in
+                return visitor.extensionDeclarations.compactMap { decl in
+                    guard visitor.typeDeclarations.contains(decl.name) else {
+                        return nil
+                    }
 
-        let susceptibleNames = Set(elements.compactMap { $0.kind != .extension ? $0.name : nil })
-
-        return elements.compactMap { element in
-            guard element.kind == .extension, susceptibleNames.contains(element.name) else {
-                return nil
+                    return ReasonedRuleViolation(position: decl.position)
+                }
             }
+            .sorted()
+            .map { makeViolation(file: file, violation: $0) }
+    }
+}
+
+private extension NoGroupingExtensionRule {
+    struct ExtensionDeclaration: Hashable {
+        let name: String
+        let position: AbsolutePosition
+    }
+
+    final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
+        private(set) var typeDeclarations = Set<String>()
+        private var typeScope: [String] = []
+        private(set) var extensionDeclarations = Set<ExtensionDeclaration>()
+
+        override var skippableDeclarations: [any DeclSyntaxProtocol.Type] {
+            [
+                ProtocolDeclSyntax.self,
+                FunctionDeclSyntax.self,
+                VariableDeclSyntax.self,
+                InitializerDeclSyntax.self,
+                SubscriptDeclSyntax.self
+            ]
```

Comment:
I think it's better to use the `allExcept(...)` method here. With that, there is a direct mapping between declarations not to skip and the implemented `visit` methods.